### PR TITLE
Convolution

### PR DESCRIPTION
**Pull request by JordanCheney**
@jklontz This is a first pass at convolution. The benchmark is showing that we are much slower then OpenCV (orders of magnitude on my machine at size > 128) any ideas for optimizations?

### CODE REVIEW

**Comment by jklontz** [comment]
For the benchmarks I try to translate the OpenCV functions as closely as possible into Likely. First the pedantic, let's rename this function to filter2D. Now we should also look more closely to see if we are implementing the same thing, understanding that the version we include in the standard library may differ to better suit our needs. In the OpenCV code it looks like the pad is always determined automatically, so lets remove it as a parameter to our function. Second, the stride is always one, so let's remove that also. filter2D includes a delta, so in fairness we should add that. After these cosmetic changes, I'd suggest stepping through the OpenCV code in your debugger. Are there any other algorithmic differences? Does OpenCV use the same approach to handling pixels that fall outside the image? Etc.

**Comment by jklontz** [comment]
Re-implemented in c655eccfe0b053200401663728252cfe0f74522e, thanks for making the first pass at it. There is still a significant slowdown but I don't have time to look into it at the moment. Probably related to how we do padding. I think it'd be better to avoid explicitly constructing the intermediate padded matrix. The OpenCV code was hard to follow w.r.t. how they handle padding.

**Comment by JordanCheney** [comment]
Thanks for doing this I got tied up with some other things. I agree
constructing the pad matrix is the wrong way to approach this. I think
there is probably a clever way of indexing the kernel such that the top
left corner of the image only multiplies the bottom right corner of the
kernel. I'll think about it and try and tackle it when I have some free time

**Comment by jklontz** [comment]
I wouldn't be afraid of trying a simple truncation `(max (min i (- size 1)) 0)` for each element access.
